extension/js/common/api/key-server/wkd.ts

```diff
@@ -36,24 +40,41 @@ export class Wkd extends Api {
       // the proper fix would be to run encodeZBase32 through background scripts
       return { pubkey: null, pgpClient: null };
     }
-    const hu = opgp.util.encodeZBase32(await opgp.crypto.hash.digest(opgp.enums.hash.sha1, Buf.fromUtfStr(user)));
-    // todo - could also search on `https://openpgpkey.{domain}/.well-known/openpgpkey/{domain}/hu/{hu}?l={user}`
-    const url = `https://${recipientDomain}/.well-known/openpgpkey/hu/${hu}?l=${encodeURIComponent(user)}`;
+    const directDomain = recipientDomain.toLowerCase();
+    const advancedDomainPrefix = (directDomain === 'localhost') ? '' : 'openpgpkey.';
+    const hu = opgp.util.encodeZBase32(await opgp.crypto.hash.digest(opgp.enums.hash.sha1, Buf.fromUtfStr(user.toLowerCase())));
+    const directHost = (typeof this.port === 'undefined') ? directDomain : `${directDomain}:${this.port}`;
+    const advancedHost = `${advancedDomainPrefix}${directHost}`;
+    const userPart = `hu/${hu}?l=${encodeURIComponent(user)}`;
+    const advancedUrl = `${this.protocol}://${advancedHost}/.well-known/openpgpkey/${directDomain}`;
+    const directUrl = `${this.protocol}://${directHost}/.well-known/openpgpkey`;
     let binary: Buf;
-    try {
-      binary = await Wkd.download(url, undefined, 4);
-    } catch (e) {
-      if (ApiErr.isNotFound(e) || ApiErr.isNetErr(e)) {
+    let validUrl: string;
+    for (const url of [advancedUrl, directUrl]) {
+      try {
+        await Wkd.download(`${url}/policy`, undefined, 4);
+        console.info(`Policy found: ${url}/policy`);
+        validUrl = url;
+        break;
+      } catch (e) {
+        if (ApiErr.isNotFound(e) || ApiErr.isNetErr(e)) {
+          continue;
+        }
+        Catch.report(`Wkd.lookupEmail error retrieving policy file ${url}/policy: ${String(e)}`);
         return { pubkey: null, pgpClient: null };
       }
-      Catch.report(`Wkd.lookupEmail err: ${String(e)}`);
+    }
+    try {
+      binary = await Wkd.download(`${validUrl!}/${userPart}`, undefined, 4);
+    } catch (e) {
+      Catch.report(`Wkd.lookupEmail error retrieving key ${validUrl!}/${userPart}: ${String(e)}`);
```

Comment:
Here should first make sure to return null/null on 404/notFound, because that's an expected behavior (not every user I try to write may have a public key even if the domain has WKD set up).

Only when it's some other type of error, then it may be worth reporting.

Reply:
Actually, the tests are failing possibly because they noticed that Catch.report was run. I think it may be ok to change these to `console.error` - so that they are only logged locally. After all, we (as a company) don't have to know about all of these reports.

Reply:
I will.
And for some reason `https://metacode.biz/.well-known/openpgpkey/policy` fails to load. I'll debug it tomorrow.

Reply:
The problem is when loading a policy file there is no `Access-Control-Allow-Origin: *` header,
whereas when getting a key there is.
Does this mean that reading policy file shouldn't be included in key extraction process, @wiktor-k ?

Reply:
I think this may be a bug in his WKD implementation. 

While spec doesn't mention CORS configuration (Wiktor tried to get it into the spec), in practice it is necessary to allow CORS. He may have forgotten to do it on the policy file.

Reply:
> He may have forgotten to do it on the policy file.

It's exactly like you've said. Additionally I'll need to put the check into WKD checker so thanks for testing that.

Reply:
The spec doesn't mention CORS but fortunately due to the checker printing warning on missing CORS a lot of people have it on. The policy issue is a interesting edge case I missed.